### PR TITLE
Fallible systems

### PR DESCRIPTION
**Pull request by NthTensor**
# Objective

Error handling in bevy is hard. See for reference https://github.com/bevyengine/bevy/issues/11562, https://github.com/bevyengine/bevy/issues/10874 and https://github.com/bevyengine/bevy/issues/12660. The goal of this PR is to make it better, by allowing users to optionally return `Result` from systems as outlined by Cart in <https://github.com/bevyengine/bevy/issues/14275#issuecomment-2223708314>.

## Solution

This PR introduces a new `ScheuleSystem` type to represent systems that can be added to schedules. Instances of this type contain either an infallible `BoxedSystem<(), ()>` or a fallible `BoxedSystem<(), Result>`. `ScheuleSystem` implements `System<In = (), Out = Result>` and replaces all uses of `BoxedSystem` in schedules. The async executor now receives a result after executing a system, which for infallible systems is always `Ok(())`. Currently it ignores this result, but more useful error handling could also be implemented.

Aliases for `Error` and `Result` have been added to the `bevy_ecs` prelude, as well as const `OK` which new users may find more friendly than `Ok(())`.

## Testing

- Currently there are not actual semantics changes that really require new tests, but I added a basic one just to make sure we don't break stuff in the future.
- The behavior of existing systems is totally unchanged, including logging.
- All of the existing systems tests pass, and I have not noticed anything strange while playing with the examples

## Showcase

The following minimal example prints "hello world" once, then completes.

```rust
use bevy::prelude::*;

fn main() {
    App::new().add_systems(Update, hello_world_system).run();
}

fn hello_world_system() -> Result {
    println!("hello world");
    Err("string")?;
    println!("goodbye world");
    OK
}
```

## Migration Guide

This change should be pretty much non-breaking, except for users who have implemented their own custom executors. Those users should use `ScheduleSystem` in place of `BoxedSystem<(), ()>` and import the `System` trait where needed. They can choose to do whatever they wish with the result.

## Current Work

+ [x] Fix tests & doc comments
+ [x] Write more tests
+ [x] Add examples
+ [X] Draft release notes

## Draft Release Notes

As of this release, systems can now return results.

First a bit of background: Bevy has hisotrically expected systems to return the empty type `()`. While this makes sense in the context of the ecs, it's at odds with how error handling is typically done in rust: returning `Result::Error` to indicate failure, and using the short-circuiting `?` operator to propagate that error up the call stack to where it can be properly handled. Users of functional languages will tell you this is called "monadic error handling".

Not being able to return `Results` from systems left bevy users with a quandry. They could add custom error handling logic to every system, or manually pipe every system into an error handler, or perhaps sidestep the issue with some combination of fallible assignents, logging, macros, and early returns. Often, users would just litter their systems with unwraps and possible panics.

While any one of these approaches might be fine for a particular user, each of them has their own drawbacks, and none makes good use of the language. Serious issues could also arrise when two different crates used by the same project made different choices about error handling.

Now, by returning results, systems can defer error handling to the application itself. It looks like this:

```rust
// Previous, handling internally
app.add_systems(my_system)
fn my_system(window: Query<&Window>) {
   let Ok(window) = query.get_single() else {
       return;
   };
   // ... do something to the window here
}

// Previous, handling externally
app.add_systems(my_system.pipe(my_error_handler))
fn my_system(window: Query<&Window>) -> Result<(), impl Error> {
   let window = query.get_single()?;
   // ... do something to the window here
   Ok(())
}

// Previous, panicking
app.add_systems(my_system)
fn my_system(window: Query<&Window>) {
   let window = query.single();
   // ... do something to the window here
}

// Now 
app.add_systems(my_system)
fn my_system(window: Query<&Window>) -> Result {
    let window = query.get_single()?;
    // ... do something to the window here
    Ok(())
}
```

There are currently some limitations. Systems must either return `()` or `Result<(), Box<dyn Error + Send + Sync + 'static>>`, with no in-between. Results are also ignored by default, and though implementing a custom handler is possible, it involves writing your own custom ecs executor (which is *not* recomended).

Systems should return errors when they cannot perform their normal behavior. In turn, errors returned to the executor while running the schedule will (eventually) be treated as unexpected. Users and library authors should prefer to return errors for anything that disrupts the normal expected behavior of a system, and should only handle expected cases internally.

We have big plans for improving error handling further:
+ Allowing users to change the error handling logic of the default executors.
+ Adding source tracking and optional backtraces to errors.
+ Possibly adding tracing-levels (Error/Warn/Info/Debug/Trace) to errors.
+ Generally making the default error logging more helpful and inteligent.
+ Adding monadic system combininators for fallible systems.
+ Possibly removing all panicking variants from our api.

### CODE REVIEW

**Comment by bushrat011899** [comment]
Even without the actual error handling benefits this provides, just having a more blessed way to use `?` in systems will be _really_ nice. I know we can just pipe the `Result` with current systems, but this will hide that bit of extra boilerplate. This also pairs nicely with making more APIs return `Result` instead of `Option`, and also makes panicking variants less important (possibly even removable TBH).

**Comment by NthTensor** [comment]
I have been informed tests are failing due to some arcane inference issue with `Never` aka `!`.

Putting this here for posterity, as I'm having to wrangle with my rust compiler to get it to update.

```rust
pub struct Never;

impl<F, Marker> IntoSystemConfigs<(Never, Marker)> for F
where
    F: IntoSystem<(), !, Marker>,
{
    fn into_configs(self) -> SystemConfigs {
        let boxed_system = Box::new(IntoSystem::into_system(self));
        SystemConfigs::new_system(ScheduleSystem::Infallible(boxed_system))
    }
}
```

**Review by bushrat011899 — APPROVED** [review]
I agree this probably needs an example, but I like the approach. Opens up the possibility of having error handlers in the future, which would resolve the _to-panic or not to-panic_ debate entirely. This also lays the groundwork for how fallibility in `Command`s could work. Really nice work!

**Comment by bushrat011899** [comment]
It was mentioned in Discord, but I'll include it here for posterity: with fallible systems getting first-class treatment, there may be room to consider removing the panicking variants of certain functions (e.g., `Query::get_entity` and `Query::entity`), since the choice of behaviour could be controlled by a system error handler. This would be a large DX win, since the "proper" methods would get the shorter names, and it'd reduce the API surface area.

**Comment by NthTensor** [comment]
> there may be room to consider removing the panicking variants of certain function

That's in line with the third point Cart proposed in https://github.com/bevyengine/bevy/issues/14275#issuecomment-2223708314. He indicated then that it was important to land all the related changes in a single release cycle, and I agree. This PR provides his (1), what (2) and (3) look like is up to @alice-i-cecile and the other designated ecs experts.

File: crates/bevy_ecs/src/result.rs
Comment (tychedelia) [inline crates/bevy_ecs/src/result.rs:7]: Since `IntoSystemConfigs` doesn't provide generic params, I'm not sure these type defaults make sense, unless there's a fast follow to support generic errors.

**Review by tychedelia — APPROVED** [review]
Amazing to see how straightforward this is, all things considered. Very excited!

File: crates/bevy_ecs/src/result.rs
Comment (NthTensor) [reply crates/bevy_ecs/src/result.rs:7]: Agreed, however I want to have these in the bevy prelude. If we put a `Result` without the generics in the prelude, it will prevent the normal result type from being used. By only adding defaults with an alias, I'm hoping we will be able to write systems that use the default `Result` and functions that use `Result<T, E>` in the same module without issue.

**Comment by teohhanhui** [comment]
> as well as const `OK` which new users may find more friendly than `Ok(())`.

Why? This just makes the code more jarring compared to the rest of the Rust ecosystem, and more cognitive load to switch between returning no results vs. returning some value.

It'd make sense if it's something useful like https://docs.rs/anyhow/latest/anyhow/fn.Ok.html

**Comment by NthTensor** [comment]
> This just makes the code more jarring compared to the rest of the Rust ecosystem

In this I am trying to defer to my understanding of Cart's preferences. He uses a const in the linked issue, and I believe has expressed that `Ok(())` is sort of confusing and cumbersome. No strong preference here from me really.

File: crates/bevy_ecs/src/result.rs
Comment (teohhanhui) [reply crates/bevy_ecs/src/result.rs:7]: There's a good precedent here: https://docs.rs/anyhow/latest/anyhow/type.Result.html

> `anyhow::Result` may be used with one or two type parameters.

**Comment by bushrat011899** [comment]
> It'd make sense if it's something useful like https://docs.rs/anyhow/latest/anyhow/fn.Ok.html

Functionally this is identical to `anyhow`'s `Ok` function. The difference is since in Bevy a `ScheduleSystem` can _only_ output an `()`, the function can instead just be a constant.

It's a problem with Rust's syntax at the end of the day. With the addition of the `Try` trait, perhaps we'll get an automatic wrapping as a bit of syntactic sugar, but that's up to the language team at the end of the day.

**Comment by alice-i-cecile** [comment]
@NthTensor let me know when you're done the outstanding tasks and take this out of draft. I'll do a final pass and then merge + create follow-up issues for this.

**Comment by alice-i-cecile** [comment]
> as well as const OK which new users may find more friendly than Ok(()).

Can you please cut this from this PR? I kinda like it, but it's pointlessly controversial here :)

File: crates/bevy_ecs/src/lib.rs
Comment (bushrat011899) [reply crates/bevy_ecs/src/lib.rs:4]: I actually prefer having the explicit allow with a `TODO`. Once 2024 releases, we will want to go back over this and potentially implement the relevant traits for systems that output `!`. If we don't, users who try to add similar `todo!` closures to their apps will fail to compile.

Comment (NthTensor) [reply crates/bevy_ecs/src/lib.rs:4]: I agree, and even if I was overruled, I'd rather not make this change as part of this PR. It's really unrelated and due only to an annoying (and imo mistaken) choice from the rust lang team.

Comment (chescock) [reply crates/bevy_ecs/src/lib.rs:4]: > Once 2024 releases, we will want to go back over this and potentially implement the relevant traits for systems that output `!`.

You could implement those traits now, right?  The 2024 edition will make those impls more useful, but you can create them before then so that you don't have to scramble after the release.  (Not suggesting that we should do that in this PR!  Just noting that we don't have to wait for the edition.)  

> I'd rather not make this change as part of this PR

Sorry, I wasn't trying to make more work for you!  Your message on Discord made it seem like you wanted to resolve the warning but couldn't, and I was just trying to offer a way to do so.  :)

Comment (bushrat011899) [reply crates/bevy_ecs/src/lib.rs:4]: > You could implement those traits now, right?

Sadly not! The `!` type isn't stabilized until Edition 2024, so we can't make the "proper" fix until then.

**Comment by NthTensor** [comment]
@alice-i-cecile requesting removal of the `M-Needs-Release-Note` label.

File: crates/bevy_ecs/src/lib.rs
Comment (chescock) [reply crates/bevy_ecs/src/lib.rs:4]: Huh, I didn't see that in the edition guide. ... Oh, I see, it's even worse than that: the plan is to stabilize it sometime *after* the edition.  

... wait, then what are we going to do when the new edition is out?  The code with the warnings won't compile on Rust 2024 without impls for `!`, but we can't add those impls until `!` is stabilized at some later date.

Comment (NthTensor) [reply crates/bevy_ecs/src/lib.rs:4]: Exactly. This is why I am a tad frustrated with the rust team. Theoretically this should not be an issue until bevy itself decides to upgrade, since each crate gets to decide its edition individually.

File: crates/bevy_ecs/src/system/schedule_system.rs
Comment (hymm) [inline crates/bevy_ecs/src/system/schedule_system.rs:15]: I’m not a fan of this approach using an enum as this is introducing a lot of branching. I’d rather see all systems returning a result and figuring out a way of wrapping systems that return ‘()’ to return Ok.

Comment (NthTensor) [reply crates/bevy_ecs/src/system/schedule_system.rs:15]: I mostly agree. Some considerations:
+ With the enum, I was thinking perhaps the compiler would know which systems are infallible and would eliminate the handler code for them.
+ When writing this, I thought maybe we might eventually want to introduce more variants.

I have since found https://github.com/bevyengine/bevy/pull/8705 and it seems to have a much more streamlined version using adapter systems. I could probably just adopt that. One issue with that version is that it mangles the origin of tracing logs, which is why it was closed. I do not believe the enum does.

Comment (NthTensor) [reply crates/bevy_ecs/src/system/schedule_system.rs:15]: If you're concerned about branching performance, we should probably just benchmark to see how bad of a regression it really is.

File: crates/bevy_ecs/src/schedule/executor/mod.rs
Comment (NthTensor) [reply crates/bevy_ecs/src/schedule/executor/mod.rs:150]: Yep they should be, good catch.

File: crates/bevy_ecs/src/lib.rs
Comment (chescock) [reply crates/bevy_ecs/src/lib.rs:4]: I don't think I follow.  What do you think the rust team should have done differently?  This looks like a standard case of code that will fail to compile on the next edition causing a warning on the current one.  It failed CI because that builds with `-Dwarnings`, but if you ignore the warning then it still builds with the old semantics.

Comment (bushrat011899) [reply crates/bevy_ecs/src/lib.rs:4]: The big issue as I see it is there'll be a period of time where the only way to accept a function that returns `!` as a parameter is to accept _all_ return types. So in this PR, we have no way of actually accepting a closure like `|| todo!()` as a system, since it returns an unnameable type in Edition 2024. Combined with the statement that they intend on back-porting this change to earlier editions makes things very complicated.

_If_ Edition 2024 launches with the never-type _and_ this change to inference, then it's fine IMO. But since it'll only launch with the inference change, we'll have this transitional period where closures are just worse.

**Comment by alice-i-cecile** [comment]
> @alice-i-cecile requesting removal of the M-Needs-Release-Note label (added in pr description because we don't have a good place for it yet).

Just like migration guides, we should keep this label around even after they're written for searchability and tooling. We might wan to rename that to be more clear though 🤔

File: crates/bevy_ecs/src/system/schedule_system.rs
Comment (hymm) [reply crates/bevy_ecs/src/system/schedule_system.rs:15]: We should check against the single threaded executor perf. I wouldn't be surprised if it isn't measurable, but I think we should still go with the other method if we can, since we know we don't need this to be dynamically settable.

Comment (NthTensor) [reply crates/bevy_ecs/src/system/schedule_system.rs:15]: If it's not a regression, and no one else feels strongly about it, I'd prefer to merge as is and refine the implementation in tree as follow-up. It will be much easier to improve various aspects of this once we have something basic working.

Comment (alice-i-cecile) [reply crates/bevy_ecs/src/system/schedule_system.rs:15]: Yeah, I don't have strong feelings about the implementation here. Getting the source-logging right is the key element to me.

File: examples/ecs/fallible_systems.rs
Comment (bushrat011899) [inline examples/ecs/fallible_systems.rs:44]: What a sight to behold. Once proper (user configurable) handlers are added in a follow-up this will be perfect. Bevy APIs can be simplified _and_ made more reliable without any loss in ergonomics (IMO). Adding `Ok(())` at the end of a system is a small price to pay that (hopefully) Rust will solve on its own (since the issue isn't specific to Bevy)

Comment (NthTensor) [reply examples/ecs/fallible_systems.rs:44]: Yeah :) this example is really just a minimal placeholder. Once we have handlers hooked up, I intend to go through and update all the examples to use this style (where it makes sense).

**Comment by BenjaminBrienen** [comment]
> > @alice-i-cecile requesting removal of the M-Needs-Release-Note label (added in pr description because we don't have a good place for it yet).
> 
> Just like migration guides, we should keep this label around even after they're written for searchability and tooling. We might wan to rename that to be more clear though 🤔 
> 

`M-#[require(ReleaseNote)]`

**Comment by NthTensor** [comment]
Alright, I added the most basic tests and examples in the world. There will be more to do there when handlers are hooked up. Ready for review.

File: examples/ecs/fallible_systems.rs
Comment (alice-i-cecile) [inline examples/ecs/fallible_systems.rs]:
```suggestion
    /// Indicate the system completed successfully:
```

File: crates/bevy_ecs/src/schedule/config.rs
Comment (alice-i-cecile) [inline crates/bevy_ecs/src/schedule/config.rs:541]: Nit: This should probably be doc hidden :)

**Review by alice-i-cecile — APPROVED** [review]
@nth this is good to go once it's merge-conflict free. I do prefer the "everything is fallible" approach by WrongShoe, but that's easily left to a follow-up refactor. Let's get the ball rolling here.

**Comment by Neo-Zhixing** [comment]
Might be a bit late on this discussion, but this particular implementation feels a bit intrusive to me. Can we instead implement IntoSystemConfigs on systems that returns `Result` so that those systems can be added to the schedule like a normal system?

**Comment by alice-i-cecile** [comment]
Thank you to everyone involved with the authoring or reviewing of this PR! This work is relatively important and needs release notes! Head over to https://github.com/bevyengine/bevy-website/issues/1967 if you'd like to help out.

**Pull request merged**
